Add GPU implem of sparse segment reduction ops

benbarsdell opened the pull request:
Adds GPU implementations of the ops: `SparseSegment[Sum|Mean|SqrtN][WithNumSegments]`.

This GPU kernel will also be used to implement the corresponding non-sparse and gradient ops in future PRs.

cc @nluehr @duncanriach

benbarsdell committed:
Add GpuSegmentedReduce helper function and test

benbarsdell committed:
Add constructors and operators to AlignedVector

benbarsdell committed:
Add GPU implem of sparse segment reduction ops

- This implementation is deterministic, vectorized, and generally faster
  than an atomic-based implementation like the existing segment
  reduction kernels (it can in fact replace them in a future commit).

benbarsdell committed:
Fix minor CI issues in PR 47974

- Reformat long line in Python test file.
- Add GPU compilation guard around GPU class specialization.

benbarsdell committed:
Attempt to fix Windows build error in PR 47974

- Pass vector types to function by reference instead of by value.
- Error was: "error C2719: formal parameter with requested alignment of
  128 won't be aligned".

benbarsdell committed:
Further attempt to fix CI build errors in PR 47974

- Avoids implicit conversion from Tinit to Tvec in kernel argument
  (the ROCm build complains about this).
- Avoids passing an aligned vector type by value to a kernel (the
  Windows build doesn't like this).

sanjoy commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -150,6 +154,423 @@ bool DisableSegmentReductionOpDeterminismExceptions() {
+      // Reduce over the segment.
+      Treducevec result = Treducevec(initial_value);
+      // Loop over the segment, reducing blockDim.y elements at a time.
+      for (Tindex yy = beg; yy < end; yy += blockDim.y) {
```
Can you please use a more descriptive name than `yy` here?

benbarsdell replied:
Renamed to `y_offset`.

sanjoy commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -150,6 +154,423 @@ bool DisableSegmentReductionOpDeterminismExceptions() {
+        Treducevec block_result =
+            x_ok && y_ok ? input_vec[input_idx] : Tvec(initial_value);
+        // Reduce over the y dimension of the block.
+        for (unsigned k = blockDim.y / 2; k > 0; k /= 2) {
+          if (x_ok && y < 2 * k) {
+            smem[y * blockDim.x + x] = block_result;
+          }
+          __syncthreads();
+          if (x_ok && y < k) {
+            block_result =
+                reduce_op(block_result, smem[(y + k) * blockDim.x + x]);
+          }
+          __syncthreads();
+        }
```
Can this be pulled into a helper function?

benbarsdell replied:
Done

benbarsdell committed:
Change SetToValue to support explicit conversion

- This avoids the need for SetToValueVectorized.

benbarsdell committed:
Fix SparseSegmentReduction num_segments

- Handles int32 or int64 num_segments.
- Changes variable names from output_rows to last_segment_id.

benbarsdell committed:
Apply GPU kernel suggestions from PR 47974

- Renames `smem` to `shared_memory`.
- Renames `beg` to `begin`.
- Renames `yy` to `y_offset`.
- Pulls GPU block reduction out into a helper device function.

benbarsdell commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -150,6 +154,423 @@ bool DisableSegmentReductionOpDeterminismExceptions() {
+// to the output type, which avoids compilation issues when Tvec is an aligned
+// vector type.
+template <typename Tvec, typename Tinit>
+__global__ void SetToValueVectorized(const int count, Tvec* __restrict__ ptr,
```
Done (modified and used the existing `SetToValue`).

benbarsdell commented on tensorflow/core/kernels/segment_reduction_ops_impl_1.cc:
```diff
@@ -66,6 +66,53 @@ bool UnsortedSegmentReductionDoValidation(OpKernel* op_kernel,
+    // Note that there is a Tnumsegments parameter on the op, but it is not
+    // plumbed through to here and so always takes its default value of int32.
+    int32 output_rows =
+        internal::SubtleMustCopy(num_segments.scalar<int32>()());
```
The original CPU code was like this. Looking at it again, I may have misinterpreted the way the type works, so the comment may not be correct.
In any case, I've now changed it to check the type as you suggest.

benbarsdell commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -150,6 +154,423 @@ bool DisableSegmentReductionOpDeterminismExceptions() {
+    const Tindex* __restrict__ indices,          // [nouter] (optional)
+    Tvec* __restrict__ output_vec) {             // [nsegments, ninner_vec]
+  GPU_DYNAMIC_SHARED_MEM_DECL(/*ALIGN = */ 16, char, smem_raw);
+  Treducevec* const smem =
```
Changed to `shared_memory`.

sanjoy replied:
Maybe add name it so that it is clearer why we're using it?

benbarsdell replied:
Renamed to `shared_partial_reduction`.

benbarsdell commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -150,6 +154,423 @@ bool DisableSegmentReductionOpDeterminismExceptions() {
+    // Grid-stride loop over segment blocks, each processing one segment.
+    for (Tsegmentids seg = blockIdx.y; seg < nsegments; seg += gridDim.y) {
+      // Load segment range.
+      const Tindex beg = segment_offsets[seg];
```
Done

benbarsdell commented on tensorflow/core/kernels/gpu_prim_helpers.h:
```diff
@@ -136,6 +136,43 @@ Status GpuInclusivePrefixSum(OpKernelContext* context, int size,
 
+template <typename InputIteratorT, typename OutputIteratorT,
+          typename OffsetIteratorT, typename ReduceOp, typename T>
+Status GpuSegmentedReduce(
```
Good question; I thought so, but I'm confirming with the CUB team.

benbarsdell replied:
I confirmed that `cub::DeviceSegmentedReduce::Reduce` is and will remain deterministic. I've added a comment noting this.

benbarsdell committed:
Add __host__ __device__ to AlignedVector operator[]

benbarsdell committed:
Address minor review comments from PR 47974

- Add comment noting that function is deterministic.
- Rename `shared_memory` to `shared_partial_reduction`.

benbarsdell commented:
Btw I haven't been able to work out or reproduce the Windows CI build failure:

```T:\tmp\nvcc_inter_files_tmp_dir\tmp9pyh0ytr\segment_reduction_ops_gpu.cu.compute_80.cudafe1.stub.c(5946): error C2719: 'unnamed-parameter': formal parameter with requested alignment of 128 won't be aligned```

If you're able to send me the temporary file `segment_reduction_ops_gpu.cu.compute_80.cudafe1.stub.c` (e.g., using `nvcc --keep`) I could try to follow up on it, but otherwise I'm not sure what to do about it.

sanjoy commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -150,6 +154,422 @@ bool DisableSegmentReductionOpDeterminismExceptions() {
+    OpKernelContext* ctx, Tindex nouter, Tindex ninner_vec,
+    Tsegmentids nsegments, ReduceOp reduce_op, Tinit initial_value,
+    Tinit empty_segment_value, bool is_mean, bool is_sqrtn,
+    const Tvec* input_vec,           // [nouter or any, ninner_vec]
```
What do you mean by "or any"?

benbarsdell replied:
Added comment explaining. When `indices` is provided, the first dim of `input_vec` is indexed indirectly and has no specific size requirement (as long as it covers the values in `indices`).

sanjoy commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -150,6 +154,422 @@ bool DisableSegmentReductionOpDeterminismExceptions() {
+    Tvec* output_vec) {              // [nsegments, ninner_vec]
+  const GPUDevice& device = ctx->eigen_gpu_device();
+
+  if (nouter == 0) {
```
Isn't this also handled in `SegmentReduceGPU`?

benbarsdell replied:
Good catch. I've removed the one in `SegmentReduceGPU` because the vectorized version should be faster.

sanjoy commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -150,6 +154,422 @@ bool DisableSegmentReductionOpDeterminismExceptions() {
+                      bool is_sqrtn, const T* input,
+                      const Tsegmentids* segment_ids, const Tindex* indices,
+                      T* output) {
+      DCHECK_EQ(ninner % vec_size, 0);  // Crash OK
+      DCHECK_EQ(reinterpret_cast<std::uintptr_t>(input) % vec_size,
+                0);  // Crash OK
+      DCHECK_EQ(reinterpret_cast<std::uintptr_t>(output) % vec_size,
+                0);  // Crash OK
```
No need for "Crash OK" for `DCHECK`.

benbarsdell replied:
Removed

benbarsdell committed:
Address more minor review comments from PR 47974

- Add parens for clarity.
- Add comment re Tvec vs. Treducevec.
- Add comment re the size of the input_vec array.
- Remove duplicate early-exit case.
- Remove unneeded "Crash OK" comments.

benbarsdell committed:
Refactor and tune SegmentReduceGPU special case

- Pulls the special ninner==1 implementation out into a separate helper
  function.
- Adds a heuristic threshold for deciding when to use the special case.
  This cub-based implementation can be much faster for large reductions,
  but can also be slower for small reductions. This is only a first pass
  at tuning the implementation; a more detailed performance study may be
  needed in future (e.g., over more of the parameter space, different
  GPUs etc.), particularly if motivated by important applications.

benbarsdell commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -150,6 +154,422 @@ bool DisableSegmentReductionOpDeterminismExceptions() {
+      Treducevec result = Treducevec(initial_value);
+      // Loop over the segment, reducing blockDim.y elements at a time.
+      for (Tindex y_offset = begin; y_offset < end; y_offset += blockDim.y) {
+        const bool y_ok = y_offset + y < end;
```
Done

benbarsdell commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -150,6 +154,422 @@ bool DisableSegmentReductionOpDeterminismExceptions() {
+  }
+}
+
+// Normalizes output_raw based on segment size and casts from Treducevec to
```
Done

benbarsdell commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -150,6 +154,422 @@ bool DisableSegmentReductionOpDeterminismExceptions() {
+  TF_RETURN_IF_ERROR(LaunchSegmentOffsetsKernel(
+      device, nouter, nsegments, segment_ids, segment_offsets_ptr));
+
+  if (ninner_vec == 1) {
```
Done.
I also added a simple heuristic based on some additional benchmarks. The performance difference appears to be very large (>4x) in some parts of the parameter space. (In general I'm treating this PR as a first pass for performance optimization, and a more detailed study can be done later).

sanjoy commented:
Ben,

> Btw I haven't been able to work out or reproduce the Windows CI build failure:
> 
> `T:\tmp\nvcc_inter_files_tmp_dir\tmp9pyh0ytr\segment_reduction_ops_gpu.cu.compute_80.cudafe1.stub.c(5946): error C2719: 'unnamed-parameter': formal parameter with requested alignment of 128 won't be aligned`
> 
> If you're able to send me the temporary file `segment_reduction_ops_gpu.cu.compute_80.cudafe1.stub.c` (e.g., using `nvcc --keep`) I could try to follow up on it, but otherwise I'm not sure what to do about it.

Sorry for not noticing this earlier, this is still breaking Windows and we have to roll back to make the build green.

Can you update the PR after `#ifdef`ing the offending code on Windows (and thus making the kernel not available on Windows)?  I don't see a better way for now, unless you're willing to reproduce this issue on a Windows VM.
